File: docs/beginner/module-05-data-sources-indexers/code-samples/python/01_azure_sql_indexer.py
#!/usr/bin/env python3
"""
Azure AI Search - SQL Database Indexer Example

This example demonstrates how to:
1. Create a data source connection to Azure SQL Database
2. Create an index for SQL data
3. Create and configure an indexer
4. Monitor indexer execution
5. Implement change tracking for incremental updates

Prerequisites:
- Azure AI Search service
- Azure SQL Database with sample data
- Required Python packages: azure-search-documents, azure-identity
"""

import os
import time
from typing import Optional
from dotenv import load_dotenv

from azure.search.documents.indexes import SearchIndexClient, SearchIndexerClient
from azure.search.documents.indexes.models import (
    SearchIndex,
    SearchField,
    SearchFieldDataType,
    SearchIndexer,
    SearchIndexerDataContainer,
    SearchIndexerDataSourceConnection,
    SearchIndexerDataSourceType,
    SqlIntegratedChangeTrackingPolicy,
    FieldMapping,
    SimpleField,
    SearchableField,
    ComplexField
)
from azure.core.credentials import AzureKeyCredential
from azure.core.exceptions import HttpResponseError

# Load environment variables
load_dotenv()

class SQLIndexerExample:
    """Example class for Azure SQL Database indexer operations."""
    
    def __init__(self):
        """Initialize the SQL indexer example with Azure credentials."""
        self.endpoint = os.getenv('SEARCH_ENDPOINT')
        self.api_key = os.getenv('SEARCH_API_KEY')
        self.sql_connection_string = os.getenv('SQL_CONNECTION_STRING')
        
        if not all([self.endpoint, self.api_key, self.sql_connection_string]):
            raise ValueError("Missing required environment variables")
        
        # Initialize clients
        credential = AzureKeyCredential(self.api_key)
        self.index_client = SearchIndexClient(self.endpoint, credential)
        self.indexer_client = SearchIndexerClient(self.endpoint, credential)
        
        # Configuration
        self.data_source_name = "sql-hotels-datasource"
        self.index_name = "hotels-sql-index"
        self.indexer_name = "hotels-sql-indexer"
        self.table_name = "Hotels"  # Adjust based on your table name
    
    def create_data_source(self) -> SearchIndexerDataSourceConnection:
        """
        Create a data source connection to Azure SQL Database.
        
        Returns:
            SearchIndexerDataSourceConnection: The created data source
        """
        print(f"Creating data source: {self.data_source_name}")
        
        # Configure data source with change tracking
        data_source = SearchIndexerDataSourceConnection(
            name=self.data_source_name,
            type=SearchIndexerDataSourceType.AZURE_SQL,
            connection_string=self.sql_connection_string,
            container=SearchIndexerDataContainer(name=self.table_name),
            data_change_detection_policy=SqlIntegratedChangeTrackingPolicy(),
            description="Hotels data from Azure SQL Database"
        )
        
        try:
            # Create or update the data source
            result = self.indexer_client.create_or_update_data_source_connection(data_source)
            print(f"✓ Data source '{self.data_source_name}' created successfully")
            return result
        except HttpResponseError as e:
            print(f"✗ Error creating data source: {e.message}")
            raise
    
    def create_index(self) -> SearchIndex:
        """
        Create a search index for hotel data.
        
        Returns:
            SearchIndex: The created search index
        """
        print(f"Creating index: {self.index_name}")
        
        # Define index fields based on SQL table schema
        fields = [
            SimpleField(name="HotelId", type=SearchFieldDataType.String, key=True),
            SearchableField(name="HotelName", type=SearchFieldDataType.String, analyzer_name="en.lucene"),
            SearchableField(name="Description", type=SearchFieldDataType.String, analyzer_name="en.lucene"),
            SimpleField(name="Category", type=SearchFieldDataType.String, filterable=True, facetable=True),
            SimpleField(name="Rating", type=SearchFieldDataType.Double, filterable=True, sortable=True, facetable=True),
            ComplexField(name="Address", fields=[
                SimpleField(name="StreetAddress", type=SearchFieldDataType.String),
                SimpleField(name="City", type=SearchFieldDataType.String, filterable=True, sortable=True, facetable=True),
                SimpleField(name="StateProvince", type=SearchFieldDataType.String, filterable=True, facetable=True),
                SimpleField(name="PostalCode", type=SearchFieldDataType.String, filterable=True),
                SimpleField(name="Country", type=SearchFieldDataType.String, filterable=True, facetable=True)
            ]),
            SimpleField(name="LastRenovationDate", type=SearchFieldDataType.DateTimeOffset, filterable=True, sortable=True)
        ]
        
        # Create the index
        index = SearchIndex(
            name=self.index_name,
            fields=fields,
            scoring_profiles=[],
            cors_options=None
        )
        
        try:
            result = self.index_client.create_or_update_index(index)
            print(f"✓ Index '{self.index_name}' created successfully")
            return result
        except HttpResponseError as e:
            print(f"✗ Error creating index: {e.message}")
            raise
    
    def create_indexer(self) -> SearchIndexer:
        """
        Create an indexer to populate the index from SQL data.
        
        Returns:
            SearchIndexer: The created indexer
        """
        print(f"Creating indexer: {self.indexer_name}")
        
        # Define field mappings if source and target field names differ
        field_mappings = [
            FieldMapping(source_field_name="HotelId", target_field_name="HotelId"),
            FieldMapping(source_field_name="HotelName", target_field_name="HotelName"),
            FieldMapping(source_field_name="Description", target_field_name="Description"),
            FieldMapping(source_field_name="Category", target_field_name="Category"),
            FieldMapping(source_field_name="Rating", target_field_name="Rating"),
            # Complex field mapping for address
            FieldMapping(source_field_name="Address", target_field_name="Address/StreetAddress"),
            FieldMapping(source_field_name="City", target_field_name="Address/City"),
            FieldMapping(source_field_name="StateProvince", target_field_name="Address/StateProvince"),
            FieldMapping(source_field_name="PostalCode", target_field_name="Address/PostalCode"),
            FieldMapping(source_field_name="Country", target_field_name="Address/Country"),
            FieldMapping(source_field_name="LastRenovationDate", target_field_name="LastRenovationDate")
        ]
        
        # Create indexer configuration
        indexer = SearchIndexer(
            name=self.indexer_name,
            data_source_name=self.data_source_name,
            target_index_name=self.index_name,
            field_mappings=field_mappings,
            description="Indexer for hotels data from SQL Database",
            parameters={
                "batchSize": 100,  # Process 100 documents at a time
                "maxFailedItems": 10,  # Allow up to 10 failed items
                "maxFailedItemsPerBatch": 5  # Allow up to 5 failures per batch
            }
        )
        
        try:
            result = self.indexer_client.create_or_update_indexer(indexer)
            print(f"✓ Indexer '{self.indexer_name}' created successfully")
            return result
        except HttpResponseError as e:
            print(f"✗ Error creating indexer: {e.message}")
            raise
    
    def run_indexer(self) -> None:
        """Run the indexer and monitor its execution."""
        print(f"Running indexer: {self.indexer_name}")
        
        try:
            # Start the indexer
            self.indexer_client.run_indexer(self.indexer_name)
            print("✓ Indexer started successfully")
            
            # Monitor execution
            self.monitor_indexer_execution()
            
        except HttpResponseError as e:
            print(f"✗ Error running indexer: {e.message}")
            raise
    
    def monitor_indexer_execution(self, timeout_seconds: int = 300) -> None:
        """
        Monitor indexer execution until completion or timeout.
        
        Args:
            timeout_seconds: Maximum time to wait for completion
        """
        print("Monitoring indexer execution...")
        start_time = time.time()
        
        while time.time() - start_time < timeout_seconds:
            try:
                status = self.indexer_client.get_indexer_status(self.indexer_name)
                
                print(f"Status: {status.status}")
                
                if status.last_result:
                    print(f"Items processed: {status.last_result.item_count}")
                    print(f"Items failed: {status.last_result.failed_item_count}")
                    print(f"Start time: {status.last_result.start_time}")
                    print(f"End time: {status.last_result.end_time}")
                    
                    # Check for errors
                    if status.last_result.errors:
                        print("Errors encountered:")
                        for error in status.last_result.errors:
                            print(f"  - {error.error_message}")
                    
                    # Check for warnings
                    if status.last_result.warnings:
                        print("Warnings:")
                        for warning in status.last_result.warnings:
                            print(f"  - {warning.message}")
                
                # Check if execution is complete
                if status.status in ["success", "error"]:
                    print(f"✓ Indexer execution completed with status: {status.status}")
                    break
                
                # Wait before checking again
                time.sleep(10)
                
            except HttpResponseError as e:
                print(f"Error checking indexer status: {e.message}")
                break
        else:
            print("⚠ Indexer monitoring timed out")
    
    def get_indexer_status(self) -> None:
        """Get and display current indexer status."""
        try:
            status = self.indexer_client.get_indexer_status(self.indexer_name)
            
            print(f"\n=== Indexer Status: {self.indexer_name} ===")
            print(f"Status: {status.status}")
            print(f"Last run status: {status.last_result.status if status.last_result else 'Never run'}")
            
            if status.last_result:
                print(f"Items processed: {status.last_result.item_count}")
                print(f"Items failed: {status.last_result.failed_item_count}")
                print(f"Execution time: {status.last_result.start_time} - {status.last_result.end_time}")
                
                if status.execution_history:
                    print(f"Total executions: {len(status.execution_history)}")
                    
        except HttpResponseError as e:
            print(f"Error getting indexer status: {e.message}")
    
    def reset_indexer(self) -> None:
        """Reset the indexer to clear its execution state."""
        try:
            self.indexer_client.reset_indexer(self.indexer_name)
            print(f"✓ Indexer '{self.indexer_name}' reset successfully")
        except HttpResponseError as e:
            print(f"✗ Error resetting indexer: {e.message}")
    
    def delete_resources(self) -> None:
        """Clean up created resources."""
        print("Cleaning up resources...")
        
        try:
            # Delete indexer
            self.indexer_client.delete_indexer(self.indexer_name)
            print(f"✓ Deleted indexer: {self.indexer_name}")
        except HttpResponseError:
            pass
        
        try:
            # Delete index
            self.index_client.delete_index(self.index_name)
            print(f"✓ Deleted index: {self.index_name}")
        except HttpResponseError:
            pass
        
        try:
            # Delete data source
            self.indexer_client.delete_data_source_connection(self.data_source_name)
            print(f"✓ Deleted data source: {self.data_source_name}")
        except HttpResponseError:
            pass
    
    def run_complete_example(self) -> None:
        """Run the complete SQL indexer example."""
        print("=== Azure AI Search SQL Indexer Example ===\n")
        
        try:
            # Step 1: Create data source
            self.create_data_source()
            
            # Step 2: Create index
            self.create_index()
            
            # Step 3: Create indexer
            self.create_indexer()
            
            # Step 4: Run indexer
            self.run_indexer()
            
            # Step 5: Check final status
            self.get_indexer_status()
            
            print("\n✓ SQL indexer example completed successfully!")
            
        except Exception as e:
            print(f"\n✗ Example failed: {str(e)}")
            raise


def main():
    """Main function to run the SQL indexer example."""
    try:
        example = SQLIndexerExample()
        example.run_complete_example()
        
        # Optionally clean up resources
        cleanup = input("\nDo you want to clean up the created resources? (y/n): ")
        if cleanup.lower() == 'y':
            example.delete_resources()
            
    except Exception as e:
        print(f"Error: {str(e)}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())

File: docs/beginner/module-05-data-sources-indexers/code-samples/javascript/01_azure_sql_indexer.js
#!/usr/bin/env node

/**
 * Azure AI Search - SQL Database Indexer Example (JavaScript)
 * 
 * This example demonstrates how to:
 * 1. Create a data source connection to Azure SQL Database
 * 2. Create an index for SQL data
 * 3. Create and configure an indexer
 * 4. Monitor indexer execution
 * 5. Implement change tracking for incremental updates
 * 
 * Prerequisites:
 * - Azure AI Search service
 * - Azure SQL Database with sample data
 * - Required Node.js packages: @azure/search-documents, @azure/identity
 */

const { SearchIndexClient, SearchIndexerClient } = require('@azure/search-documents');
const { AzureKeyCredential } = require('@azure/core-auth');
const { RestError } = require('@azure/core-rest-pipeline');
require('dotenv').config();

class SQLIndexerExample {
    constructor() {
        // Load configuration from environment variables
        this.endpoint = process.env.SEARCH_ENDPOINT;
        this.apiKey = process.env.SEARCH_API_KEY;
        this.sqlConnectionString = process.env.SQL_CONNECTION_STRING;
        
        if (!this.endpoint || !this.apiKey || !this.sqlConnectionString) {
            throw new Error('Missing required environment variables');
        }
        
        // Initialize clients
        const credential = new AzureKeyCredential(this.apiKey);
        this.indexClient = new SearchIndexClient(this.endpoint, credential);
        this.indexerClient = new SearchIndexerClient(this.endpoint, credential);
        
        // Configuration
        this.dataSourceName = 'sql-hotels-datasource';
        this.indexName = 'hotels-sql-index';
        this.indexerName = 'hotels-sql-indexer';
        this.tableName = 'Hotels';
    }
    
    /**
     * Create a data source connection to Azure SQL Database
     */
    async createDataSource() {
        console.log(`Creating data source: ${this.dataSourceName}`);
        
        const dataSource = {
            name: this.dataSourceName,
            type: 'azuresql',
            connectionString: this.sqlConnectionString,
            container: { name: this.tableName },
            dataChangeDetectionPolicy: {
                '@odata.type': '#Microsoft.Azure.Search.SqlIntegratedChangeTrackingPolicy'
            },
            description: 'Hotels data from Azure SQL Database'
        };
        
        try {
            const result = await this.indexerClient.createOrUpdateDataSourceConnection(dataSource);
            console.log(`✅ Data source '${this.dataSourceName}' created successfully`);
            return result;
        } catch (error) {
            console.error(`❌ Error creating data source: ${error.message}`);
            throw error;
        }
    }
    
    /**
     * Create a search index for hotel data
     */
    async createIndex() {
        console.log(`Creating index: ${this.indexName}`);
        
        const index = {
            name: this.indexName,
            fields: [
                {
                    name: 'HotelId',
                    type: 'Edm.String',
                    key: true,
                    retrievable: true
                },
                {
                    name: 'HotelName',
                    type: 'Edm.String',
                    searchable: true,
                    retrievable: true,
                    analyzer: 'en.lucene'
                },
                {
                    name: 'Description',
                    type: 'Edm.String',
                    searchable: true,
                    retrievable: true,
                    analyzer: 'en.lucene'
                },
                {
                    name: 'Category',
                    type: 'Edm.String',
                    filterable: true,
                    facetable: true,
                    retrievable: true
                },
                {
                    name: 'Rating',
                    type: 'Edm.Double',
                    filterable: true,
                    sortable: true,
                    facetable: true,
                    retrievable: true
                },
                {
                    name: 'Address',
                    type: 'Edm.ComplexType',
                    fields: [
                        {
                            name: 'StreetAddress',
                            type: 'Edm.String',
                            retrievable: true
                        },
                        {
                            name: 'City',
                            type: 'Edm.String',
                            filterable: true,
                            sortable: true,
                            facetable: true,
                            retrievable: true
                        },
                        {
                            name: 'StateProvince',
                            type: 'Edm.String',
                            filterable: true,
                            facetable: true,
                            retrievable: true
                        },
                        {
                            name: 'PostalCode',
                            type: 'Edm.String',
                            filterable: true,
                            retrievable: true
                        },
                        {
                            name: 'Country',
                            type: 'Edm.String',
                            filterable: true,
                            facetable: true,
                            retrievable: true
                        }
                    ]
                },
                {
                    name: 'LastRenovationDate',
                    type: 'Edm.DateTimeOffset',
                    filterable: true,
                    sortable: true,
                    retrievable: true
                }
            ]
        };
        
        try {
            const result = await this.indexClient.createOrUpdateIndex(index);
            console.log(`✅ Index '${this.indexName}' created successfully`);
            return result;
        } catch (error) {
            console.error(`❌ Error creating index: ${error.message}`);
            throw error;
        }
    }
    
    /**
     * Create an indexer to populate the index from SQL data
     */
    async createIndexer() {
        console.log(`Creating indexer: ${this.indexerName}`);
        
        const indexer = {
            name: this.indexerName,
            dataSourceName: this.dataSourceName,
            targetIndexName: this.indexName,
            fieldMappings: [
                { sourceFieldName: 'HotelId', targetFieldName: 'HotelId' },
                { sourceFieldName: 'HotelName', targetFieldName: 'HotelName' },
                { sourceFieldName: 'Description', targetFieldName: 'Description' },
                { sourceFieldName: 'Category', targetFieldName: 'Category' },
                { sourceFieldName: 'Rating', targetFieldName: 'Rating' },
                { sourceFieldName: 'Address', targetFieldName: 'Address/StreetAddress' },
                { sourceFieldName: 'City', targetFieldName: 'Address/City' },
                { sourceFieldName: 'StateProvince', targetFieldName: 'Address/StateProvince' },
                { sourceFieldName: 'PostalCode', targetFieldName: 'Address/PostalCode' },
                { sourceFieldName: 'Country', targetFieldName: 'Address/Country' },
                { sourceFieldName: 'LastRenovationDate', targetFieldName: 'LastRenovationDate' }
            ],
            description: 'Indexer for hotels data from SQL Database',
            parameters: {
                batchSize: 100,
                maxFailedItems: 10,
                maxFailedItemsPerBatch: 5
            }
        };
        
        try {
            const result = await this.indexerClient.createOrUpdateIndexer(indexer);
            console.log(`✅ Indexer '${this.indexerName}' created successfully`);
            return result;
        } catch (error) {
            console.error(`❌ Error creating indexer: ${error.message}`);
            throw error;
        }
    }
    
    /**
     * Run the indexer and monitor its execution
     */
    async runIndexer() {
        console.log(`Running indexer: ${this.indexerName}`);
        
        try {
            // Start the indexer
            await this.indexerClient.runIndexer(this.indexerName);
            console.log('✅ Indexer started successfully');
            
            // Monitor execution
            await this.monitorIndexerExecution();
            
        } catch (error) {
            console.error(`❌ Error running indexer: ${error.message}`);
            throw error;
        }
    }
    
    /**
     * Monitor indexer execution until completion or timeout
     */
    async monitorIndexerExecution(timeoutSeconds = 300) {
        console.log('Monitoring indexer execution...');
        const startTime = Date.now();
        
        while (Date.now() - startTime < timeoutSeconds * 1000) {
            try {
                const status = await this.indexerClient.getIndexerStatus(this.indexerName);
                
                console.log(`Status: ${status.status}`);
                
                if (status.lastResult) {
                    console.log(`Items processed: ${status.lastResult.itemCount || 0}`);
                    console.log(`Items failed: ${status.lastResult.failedItemCount || 0}`);
                    console.log(`Start time: ${status.lastResult.startTime}`);
                    console.log(`End time: ${status.lastResult.endTime}`);
                    
                    // Check for errors
                    if (status.lastResult.errors && status.lastResult.errors.length > 0) {
                        console.log('Errors encountered:');
                        status.lastResult.errors.forEach(error => {
                            console.log(`  - ${error.errorMessage}`);
                        });
                    }
                    
                    // Check for warnings
                    if (status.lastResult.warnings && status.lastResult.warnings.length > 0) {
                        console.log('Warnings:');
                        status.lastResult.warnings.forEach(warning => {
                            console.log(`  - ${warning.message}`);
                        });
                    }
                }
                
                // Check if execution is complete
                if (status.status === 'success' || status.status === 'error') {
                    console.log(`✅ Indexer execution completed with status: ${status.status}`);
                    break;
                }
                
                // Wait before checking again
                await new Promise(resolve => setTimeout(resolve, 10000));
                
            } catch (error) {
                console.error(`Error checking indexer status: ${error.message}`);
                break;
            }
        }
    }
    
    /**
     * Get and display current indexer status
     */
    async getIndexerStatus() {
        try {
            const status = await this.indexerClient.getIndexerStatus(this.indexerName);
            
            console.log(`\n=== Indexer Status: ${this.indexerName} ===`);
            console.log(`Status: ${status.status}`);
            console.log(`Last run status: ${status.lastResult ? status.lastResult.status : 'Never run'}`);
            
            if (status.lastResult) {
                console.log(`Items processed: ${status.lastResult.itemCount || 0}`);
                console.log(`Items failed: ${status.lastResult.failedItemCount || 0}`);
                console.log(`Execution time: ${status.lastResult.startTime} - ${status.lastResult.endTime}`);
                
                if (status.executionHistory) {
                    console.log(`Total executions: ${status.executionHistory.length}`);
                }
            }
            
        } catch (error) {
            console.error(`Error getting indexer status: ${error.message}`);
        }
    }
    
    /**
     * Reset the indexer to clear its execution state
     */
    async resetIndexer() {
        try {
            await this.indexerClient.resetIndexer(this.indexerName);
            console.log(`✅ Indexer '${this.indexerName}' reset successfully`);
        } catch (error) {
            console.error(`❌ Error resetting indexer: ${error.message}`);
        }
    }
    
    /**
     * Clean up created resources
     */
    async deleteResources() {
        console.log('Cleaning up resources...');
        
        try {
            await this.indexerClient.deleteIndexer(this.indexerName);
            console.log(`✅ Deleted indexer: ${this.indexerName}`);
        } catch (error) {
            // Ignore errors during cleanup
        }
        
        try {
            await this.indexClient.deleteIndex(this.indexName);
            console.log(`✅ Deleted index: ${this.indexName}`);
        } catch (error) {
            // Ignore errors during cleanup
        }
        
        try {
            await this.indexerClient.deleteDataSourceConnection(this.dataSourceName);
            console.log(`✅ Deleted data source: ${this.dataSourceName}`);
        } catch (error) {
            // Ignore errors during cleanup
        }
    }
    
    /**
     * Run the complete SQL indexer example
     */
    async runCompleteExample() {
        console.log('=== Azure AI Search SQL Indexer Example ===\n');
        
        try {
            // Step 1: Create data source
            await this.createDataSource();
            
            // Step 2: Create index
            await this.createIndex();
            
            // Step 3: Create indexer
            await this.createIndexer();
            
            // Step 4: Run indexer
            await this.runIndexer();
            
            // Step 5: Check final status
            await this.getIndexerStatus();
            
            console.log('\n✅ SQL indexer example completed successfully!');
            
        } catch (error) {
            console.error(`\n❌ Example failed: ${error.message}`);
            throw error;
        }
    }
}

// Main execution
async function main() {
    try {
        const example = new SQLIndexerExample();
        await example.runCompleteExample();
        
        // Optionally clean up resources
        const readline = require('readline').createInterface({
            input: process.stdin,
            output: process.stdout
        });
        
        readline.question('\nDo you want to clean up the created resources? (y/n): ', async (answer) => {
            if (answer.toLowerCase() === 'y') {
                await example.deleteResources();
            }
            readline.close();
            process.exit(0);
        });
        
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

// Run the example if this file is executed directly
if (require.main === module) {
    main();
}

module.exports = SQLIndexerExample;